modules/autotagging-commons/common/src/main/java/org/opensearch/rule/service/IndexStoredRulePersistenceService.java

```diff
@@ -43,42 +55,152 @@ public class IndexStoredRulePersistenceService implements RulePersistenceService
      * This service handles persistence and retrieval of stored rules within an OpenSearch index.
      * @param indexName - The name of the OpenSearch index where the rules are stored.
      * @param client - The OpenSearch client used to interact with the OpenSearch cluster.
+     * @param clusterService
      * @param maxRulesPerPage - The maximum number of rules that can be returned in a single get request.
      * @param parser
      * @param queryBuilder
      */
     public IndexStoredRulePersistenceService(
         String indexName,
         Client client,
+        ClusterService clusterService,
         int maxRulesPerPage,
         RuleEntityParser parser,
         RuleQueryMapper<QueryBuilder> queryBuilder
     ) {
         this.indexName = indexName;
         this.client = client;
+        this.clusterService = clusterService;
         this.maxRulesPerPage = maxRulesPerPage;
         this.parser = parser;
         this.queryBuilder = queryBuilder;
     }
 
+    /**
+     * Entry point for the create rule API logic in persistence service.
+     * It ensures the index exists, validates for duplicate rules, and persists the new rule.
+     * @param request  The CreateRuleRequest
+     * @param listener ActionListener for CreateRuleResponse
+     */
+    public void createRule(CreateRuleRequest request, ActionListener<CreateRuleResponse> listener) {
+        try (ThreadContext.StoredContext ctx = getContext()) {
+            if (!clusterService.state().metadata().hasIndex(indexName)) {
+                createIndex(listener, request.getRule());
+            } else {
+                validateAndPersist(request.getRule(), listener);
+            }
+        }
+    }
+
+    /**
+     * Creates the system index, then validates and persists the given rule.
+     * @param listener - ActionListener for CreateRuleResponse
+     * @param rule - the rule to validate and persist
+     */
+    private void createIndex(ActionListener<CreateRuleResponse> listener, Rule rule) {
+        final CreateIndexRequest request = new CreateIndexRequest(indexName).settings(indexSettings);
+        client.admin().indices().create(request, new ActionListener<>() {
+            @Override
+            public void onResponse(CreateIndexResponse response) {
+                if (!response.isAcknowledged()) {
+                    logger.error("Index creation not acknowledged: {}", indexName);
+                    listener.onFailure(new IllegalStateException(indexName + " index creation failed and rule cannot be persisted"));
+                } else {
+                    validateAndPersist(rule, listener);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                Throwable cause = ExceptionsHelper.unwrapCause(e);
+                if (cause instanceof ResourceAlreadyExistsException) {
+                    validateAndPersist(rule, listener);
+                } else {
```

Comment:
I am assuming this case only occurs if 2 operations try to create the index concurrently. Only 1 of them will succeed?

Reply:
Have we added tests for all these scenarios?